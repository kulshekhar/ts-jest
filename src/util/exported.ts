export { mocked } from './testing'
export { createJestPreset } from '../config/create-jest-preset'
export { pathsToModuleNameMapper } from '../config/paths-to-module-name-mapper'
export { projectsToModuleNameMapper } from '../config/projects-to-module-name-mapper'
export { tsConfigToModuleNameMapper } from '../config/tsconfig-to-module-name-mapper'
