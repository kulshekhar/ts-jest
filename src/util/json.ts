/* eslint-disable no-redeclare */
import stableStringify = require('fast-json-stable-stringify')

const UNDEFINED = 'undefined'

/**
 * @internal
 */
export function stringify(input: any): string {
  return input === undefined ? UNDEFINED : stableStringify(input)
}

/**
 * @internal
 */
export function parse(input: string): any {
  return input === UNDEFINED ? undefined : JSON.parse(input)
}

interface NormalizeOptions {
  parse?: (input: string) => any
}
/**
 * @internal
 */
export function normalize(input: string, { parse: parser = parse }: NormalizeOptions = {}): string {
  let result: string | undefined
  if (normalize.cache.has(input)) {
    result = normalize.cache.get(input)
  } else {
    const data = parser(input)
    result = stringify(data)
    if (result === input) result = undefined
    normalize.cache.set(input, result)
  }
  return result === undefined ? input : result
}

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace normalize {
  export const cache = new Map<string, string | undefined>()
}
