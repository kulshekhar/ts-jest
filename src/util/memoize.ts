const cacheProp = Symbol.for('[memoize]') as any

/**
 * @internal
 */
export function Memoize(keyBuilder?: (...args: any[]) => any) {
  /* eslint-disable-next-line  @typescript-eslint/ban-types */
  return (_: object, propertyKey: string, descriptor: TypedPropertyDescriptor<any>): void => {
    if (descriptor.value != null) {
      descriptor.value = memoize(propertyKey, descriptor.value, keyBuilder || ((v: any) => v))
    } else if (descriptor.get != null) {
      descriptor.get = memoize(propertyKey, descriptor.get, keyBuilder || (() => propertyKey))
    }
  }
}

// See https://github.com/microsoft/TypeScript/issues/1863#issuecomment-579541944
/* eslint-disable-next-line  @typescript-eslint/ban-types */
function ensureCache<T extends Record<string, unknown> & { [key: string]: any }>(
  target: T,
  reset = false,
): { [key in keyof T]?: Map<any, any> } {
  if (reset || !target[cacheProp]) {
    Object.defineProperty(target, cacheProp, {
      value: Object.create(null),
      configurable: true,
    })
  }

  return target[cacheProp]
}

// See https://github.com/microsoft/TypeScript/issues/1863#issuecomment-579541944
/* eslint-disable-next-line  @typescript-eslint/ban-types */
function ensureChildCache<T extends Record<string, unknown> & { [key: string]: any }>(
  target: T,
  key: keyof T,
  reset = false,
): Map<any, any> {
  const dict = ensureCache(target)
  if (reset || !dict[key]) {
    dict[key] = new Map<any, any>()
  }

  return dict[key] as Map<any, any>
}

function memoize(
  namespace: string,
  func: (...args: any[]) => any,
  keyBuilder: (...args: any[]) => any,
): (...args: any[]) => any {
  return function (this: any, ...args: any[]): any {
    const cache = ensureChildCache(this, namespace)
    const key = keyBuilder.apply(this, args)
    if (cache.has(key)) return cache.get(key)
    const res: any = func.apply(this, args)
    cache.set(key, res)
    return res
  }
}
