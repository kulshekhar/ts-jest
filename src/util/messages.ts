// tslint:disable:max-line-length

/**
 * @internal
 */
export enum Errors {
  LoadingModuleFailed = 'Loading module {{module}} failed with error: {{error}}',
  UnableToLoadOneModule = 'Unable to load the module {{module}}. {{reason}} To fix it:\n{{fix}}',
  UnableToLoadAnyModule = 'Unable to load any of these modules: {{module}}. {{reason}}. To fix it:\n{{fix}}',
  TypesUnavailableWithoutTypeCheck = 'Type information is unavailable with "isolatedModules"',
  UnableToRequireDefinitionFile = 'Unable to require `.d.ts` file.\nThis is usually the result of a faulty configuration or import. Make sure there is a `.js`, `.json` or another executable extension available alongside `{{file}}`.',
  FileNotFound = 'File not found: {{inputPath}} (resolved as: {{resolvedPath}})',
  UntestedDependencyVersion = "Version {{actualVersion}} of {{module}} installed has not been tested with ts-jest. If you're experiencing issues, consider using a supported version ({{expectedVersion}}). Please do not report issues in ts-jest if you are using unsupported versions.",
  MissingDependency = "Module {{module}} is not installed. If you're experiencing issues, consider installing a supported version ({{expectedVersion}}).",
  UnableToCompileTypeScript = 'TypeScript diagnostics ({{help}}):\n{{diagnostics}}',
  NotMappingMultiStarPath = 'Not mapping "{{path}}" because it has more than one star (`*`).',
  NotMappingPathWithEmptyMap = 'Not mapping "{{path}}" because it has no target.',
  MappingOnlyFirstTargetOfPath = 'Mapping only to first target of "{{path}}" because it has more than one ({{count}}).',
  GotJsFileButAllowJsFalse = 'Got a `.js` file to compile while `allowJs` option is not set to `true` (file: {{path}}). To fix this:\n  - if you want TypeScript to process JS files, set `allowJs` to `true` in your TypeScript config (usually tsconfig.json)\n  - if you do not want TypeScript to process your `.js` files, in your Jest config change the `transform` key which value is `ts-jest` so that it does not match `.js` files anymore',
  GotUnknownFileTypeWithoutBabel = 'Got a unknown file type to compile (file: {{path}}). To fix this, in your Jest config change the `transform` key which value is `ts-jest` so that it does not match this kind of files anymore.',
  GotUnknownFileTypeWithBabel = 'Got a unknown file type to compile (file: {{path}}). To fix this, in your Jest config change the `transform` key which value is `ts-jest` so that it does not match this kind of files anymore. If you still want Babel to process it, add another entry to the `transform` option with value `babel-jest` which key matches this type of files.',
  ConfigNoModuleInterop = 'If you have issues related to imports, you should consider setting `esModuleInterop` to `true` in your TypeScript configuration file (usually `tsconfig.json`). See https://blogs.msdn.microsoft.com/typescript/2018/01/31/announcing-typescript-2-7/#easier-ecmascript-module-interoperability for more information.',
  UnableToFindProjectRoot = 'Unable to find the root of the project where ts-jest has been installed.',
  UnableToResolveJestConfig = 'Unable to resolve jest-config. Ensure Jest is properly installed.',
}

/**
 * @internal
 */
export enum Helps {
  FixMissingModule = '{{label}}: `npm i -D {{module}}` (or `yarn add --dev {{module}}`)',
  IgnoreDiagnosticCode = 'customize using `[jest-config].globals.ts-jest.diagnostics` option',
  MigrateConfigUsingCLI = 'Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.',
}

/**
 * @internal
 */
export enum Deprecateds {
  EnvVar = 'Using env. var "{{old}}" is deprecated, use "{{new}}" instead.',
  ConfigOption = '"[jest-config].{{oldPath}}" is deprecated, use "[jest-config].{{newPath}}" instead.',
  ConfigOptionWithNote = '"[jest-config].{{oldPath}}" is deprecated, use "[jest-config].{{newPath}}" instead.\n    â†³ {{note}}',
  ConfigOptionUseBabelRcNote = 'See `babel-jest` related issue: https://github.com/facebook/jest/issues/3845',
  HelperMovedToUtils = "The `{{helper}}` helper has been moved to `ts-jest/utils`. Use `import { {{helper}} } from 'ts-jest/utils'` instead.",
}

/**
 * @internal
 */
export enum ImportReasons {
  TsJest = 'Using "ts-jest" requires this package to be installed.',
  BabelJest = 'Using "babel-jest" requires this package to be installed.',
}

/**
 * @internal
 */
export function interpolate(msg: string, vars: Record<string, any> = {}): string {
  return msg.replace(/\{\{([^\}]+)\}\}/g, (_, key) => (key in vars ? vars[key] : _))
}
