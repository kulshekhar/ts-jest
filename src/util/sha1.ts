import { createHash } from 'crypto'

/**
 * @internal
 */
// stores hashes made out of only one argument being a string
export const cache: { [key: string]: string } = Object.create(null)

type DataItem = string | Buffer

/**
 * @internal
 */
export function sha1(...data: DataItem[]): string {
  const canCache = data.length === 1 && typeof data[0] === 'string'
  // caching
  let cacheKey!: string
  if (canCache) {
    cacheKey = data[0] as string
    if (cacheKey in cache) {
      return cache[cacheKey]
    }
  }

  // we use SHA1 because it's the fastest provided by node
  // and we are not concerned about security here
  const hash = createHash('sha1')
  data.forEach((item) => {
    if (typeof item === 'string') hash.update(item, 'utf8')
    else hash.update(item)
  })
  const res = hash.digest('hex').toString()

  if (canCache) {
    cache[cacheKey] = res
  }

  return res
}
