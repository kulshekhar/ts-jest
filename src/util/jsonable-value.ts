import { stringify } from './json'

/**
 * @internal
 */
export class JsonableValue<V = any> {
  private _serialized!: string
  private _value!: V

  constructor(value: V) {
    this.value = value
  }

  set value(value: V) {
    this._value = value
    this._serialized = stringify(value)
  }

  get value(): V {
    return this._value
  }

  get serialized(): string {
    return this._serialized
  }

  valueOf(): V {
    return this._value
  }

  toString(): string {
    return this._serialized
  }
}
