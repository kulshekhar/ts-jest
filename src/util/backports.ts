import { LogContexts, Logger } from 'bs-logger'

import { Deprecateds, Helps, interpolate } from './messages'

const context = { [LogContexts.namespace]: 'backports' }

export const backportJestConfig = <T extends jest.InitialOptions | jest.ProjectConfig>(
  logger: Logger,
  config: T,
): T => {
  logger.debug({ ...context, config }, 'backporting config')

  const { globals = {} } = (config || {}) as any
  const { 'ts-jest': tsJest = {} } = globals as any
  const mergeTsJest: any = {}
  let hadWarnings = false
  const warnConfig = (oldPath: string, newPath: string, note?: string) => {
    hadWarnings = true
    logger.warn(
      context,
      interpolate(note ? Deprecateds.ConfigOptionWithNote : Deprecateds.ConfigOption, {
        oldPath,
        newPath,
        note,
      }),
    )
  }

  if ('__TS_CONFIG__' in globals) {
    warnConfig('globals.__TS_CONFIG__', 'globals.ts-jest.tsConfig')
    if (typeof globals.__TS_CONFIG__ === 'object') {
      mergeTsJest.tsConfig = globals.__TS_CONFIG__
    }
    delete globals.__TS_CONFIG__
  }

  if ('__TRANSFORM_HTML__' in globals) {
    warnConfig('globals.__TRANSFORM_HTML__', 'globals.ts-jest.stringifyContentPathRegex')
    if (globals.__TRANSFORM_HTML__) {
      mergeTsJest.stringifyContentPathRegex = '\\.html?$'
    }
    delete globals.__TRANSFORM_HTML__
  }

  if ('typeCheck' in tsJest) {
    warnConfig('globals.ts-jest.typeCheck', 'globals.ts-jest.isolatedModules')
    mergeTsJest.isolatedModules = !tsJest.typeCheck
    delete tsJest.typeCheck
  }

  if ('tsConfigFile' in tsJest) {
    warnConfig('globals.ts-jest.tsConfigFile', 'globals.ts-jest.tsConfig')
    if (tsJest.tsConfigFile) {
      mergeTsJest.tsConfig = tsJest.tsConfigFile
    }
    delete tsJest.tsConfigFile
  }

  if ('enableTsDiagnostics' in tsJest) {
    warnConfig('globals.ts-jest.enableTsDiagnostics', 'globals.ts-jest.diagnostics')
    if (tsJest.enableTsDiagnostics) {
      mergeTsJest.diagnostics = { warnOnly: true }
      if (typeof tsJest.enableTsDiagnostics === 'string') mergeTsJest.diagnostics.pathRegex = tsJest.enableTsDiagnostics
    } else {
      mergeTsJest.diagnostics = false
    }
    delete tsJest.enableTsDiagnostics
  }

  if ('useBabelrc' in tsJest) {
    warnConfig('globals.ts-jest.useBabelrc', 'globals.ts-jest.babelConfig', Deprecateds.ConfigOptionUseBabelRcNote)
    if (tsJest.useBabelrc != null) {
      mergeTsJest.babelConfig = tsJest.useBabelrc ? true : {}
    }
    delete tsJest.useBabelrc
  }

  // if ('babelConfig' in tsJest) {
  //   warnConfig('globals.ts-jest.babelConfig', 'globals.ts-jest.babelConfig')
  //   if (tsJest.babelConfig != null) {
  //     mergeTsJest.babelConfig = tsJest.babelConfig
  //   }
  //   delete tsJest.babelConfig
  // }

  if ('skipBabel' in tsJest) {
    warnConfig('globals.ts-jest.skipBabel', 'globals.ts-jest.babelConfig')
    if (tsJest.skipBabel === false && !mergeTsJest.babelConfig) {
      mergeTsJest.babelConfig = true
    }
    delete tsJest.skipBabel
  }

  // if we had some warnings we can inform the user about the CLI tool
  if (hadWarnings) {
    logger.warn(context, Helps.MigrateConfigUsingCLI)
  }

  return {
    ...(config as any),
    globals: {
      ...globals,
      'ts-jest': {
        ...mergeTsJest,
        ...tsJest,
      },
    },
  }
}

export const backportTsJestDebugEnvVar = (logger: Logger) => {
  if ('TS_JEST_DEBUG' in process.env) {
    const shouldLog = !/^\s*(?:0|f(?:alse)?|no?|disabled?|off|)\s*$/i.test(process.env.TS_JEST_DEBUG || '')
    delete process.env.TS_JEST_DEBUG
    if (shouldLog) {
      process.env.TS_JEST_LOG = `ts-jest.log,stderr:warn`
    }
    logger.warn(
      context,
      interpolate(Deprecateds.EnvVar, {
        old: 'TS_JEST_DEBUG',
        new: 'TS_JEST_LOG',
      }),
    )
  }
}
