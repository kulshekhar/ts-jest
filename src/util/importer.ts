import { ModulePatcher, TBabelCore, TBabelJest, TTypeScript } from '../types'

import { rootLogger } from './logger'
import { Memoize } from './memoize'
import { Errors, Helps, ImportReasons, interpolate } from './messages'
import { VersionCheckers } from './version-checkers'

const logger = rootLogger.child({ namespace: 'Importer' })

// When adding an optional dependency which has another reason, add the reason in ImportReasons, and
// create a new method in Importer. Thus uses the importer.yourMethod(ImportReasons.TheReason)
// in the relevant code, so that the user knows why it needs it and how to install it in the
// case it can't import.

interface ImportOptions {
  alternatives?: string[]
  installTip?: string | Array<{ module: string; label: string }>
}

const passThru = (action: () => void) => (input: any) => {
  action()
  return input
}

/**
 * @internal
 */
export class Importer {
  @Memoize()
  static get instance() {
    logger.debug('creating Importer singleton')
    // here we can define patches to apply to modules.
    // it could be fixes that are not deployed, or
    // abstractions so that multiple versions work the same
    return new Importer({
      '@babel/core': [passThru(VersionCheckers.babelCore.warn)],
      'babel-jest': [passThru(VersionCheckers.babelJest.warn)],
      typescript: [passThru(VersionCheckers.typescript.warn)],
      jest: [passThru(VersionCheckers.jest.warn)],
    })
  }

  constructor(protected _patches: { [moduleName: string]: ModulePatcher[] } = {}) {}

  babelJest(why: ImportReasons): TBabelJest {
    return this._import(why, 'babel-jest')
  }

  babelCore(why: ImportReasons): TBabelCore {
    return this._import(why, '@babel/core')
  }

  typescript(why: ImportReasons, which: string): TTypeScript {
    return this._import(why, which)
  }

  @Memoize((...args: string[]) => args.join(':'))
  tryThese(moduleName: string, ...fallbacks: string[]) {
    let name: string
    let loaded: RequireResult<true> | undefined
    const tries = [moduleName, ...fallbacks]
    // tslint:disable-next-line:no-conditional-assignment
    while ((name = tries.shift() as string) !== undefined) {
      const req = requireWrapper(name)

      // remove exports from what we're going to log
      const contextReq: any = { ...req }
      delete contextReq.exports

      if (req.exists) {
        // module exists
        loaded = req as RequireResult<true>
        if (loaded.error) {
          // require-ing it failed
          logger.error({ requireResult: contextReq }, `failed loading module '${name}'`, loaded.error.message)
        } else {
          // it has been loaded, let's patch it
          logger.debug({ requireResult: contextReq }, 'loaded module', name)
          loaded.exports = this._patch(name, loaded.exports)
        }
        break
      } else {
        // module does not exists in the path
        logger.debug({ requireResult: contextReq }, `module '${name}' not found`)
        continue
      }
    }

    // return the loaded one, could be one that has been loaded, or one which has failed during load
    // but not one which does not exists
    return loaded
  }

  tryTheseOr<T>(moduleNames: [string, ...string[]] | string, missingResult: T, allowLoadError?: boolean): T
  tryTheseOr<T>(moduleNames: [string, ...string[]] | string, missingResult?: T, allowLoadError?: boolean): T | undefined
  tryTheseOr<T>(moduleNames: [string, ...string[]] | string, missingResult?: T, allowLoadError = false): T | undefined {
    const args: [string, ...string[]] = Array.isArray(moduleNames) ? moduleNames : [moduleNames]
    const result = this.tryThese(...args)
    if (!result) return missingResult
    if (!result.error) return result.exports as T
    if (allowLoadError) return missingResult
    throw result.error
  }

  @Memoize(name => name)
  protected _patch<T>(name: string, unpatched: T): T {
    if (name in this._patches) {
      logger.debug('patching', name)
      return this._patches[name].reduce((mod, patcher) => patcher(mod), unpatched)
    }
    return unpatched
  }

  protected _import(
    why: string,
    moduleName: string,
    { alternatives = [], installTip = moduleName }: ImportOptions = {},
  ): any {
    // try to load any of the alternative after trying main one
    const res = this.tryThese(moduleName, ...alternatives)
    // if we could load one, return it
    if (res && res.exists) {
      if (!res.error) return res.exports
      // it could not load because of a failure while importing, but it exists
      throw new Error(interpolate(Errors.LoadingModuleFailed, { module: res.given, error: res.error.message }))
    }

    // if it couldn't load, build a nice error message so the user can fix it by himself
    const msg = alternatives.length ? Errors.UnableToLoadAnyModule : Errors.UnableToLoadOneModule
    const loadModule = [moduleName, ...alternatives].map(m => `"${m}"`).join(', ')
    if (typeof installTip === 'string') {
      installTip = [{ module: installTip, label: `install "${installTip}"` }]
    }
    const fix = installTip
      .map(tip => {
        return `    ${installTip.length === 1 ? '↳' : '•'} ${interpolate(Helps.FixMissingModule, tip)}`
      })
      .join('\n')

    throw new Error(
      interpolate(msg, {
        module: loadModule,
        reason: why,
        fix,
      }),
    )
  }
}

/**
 * @internal
 */
export const importer = Importer.instance

/**
 * @internal
 */
export interface RequireResult<E = boolean> {
  exists: E
  given: string
  path?: string
  exports?: any
  error?: Error
}

function requireWrapper(moduleName: string): RequireResult {
  let path: string
  let exists = false
  try {
    path = resolveModule(moduleName)
    exists = true
  } catch (error) {
    return { error, exists, given: moduleName }
  }
  const result: RequireResult = { exists, path, given: moduleName }
  try {
    result.exports = requireModule(moduleName)
  } catch (error) {
    result.error = error
  }
  return result
}

let requireModule = (mod: string) => require(mod)
let resolveModule = (mod: string) => require.resolve(mod)

/**
 * @internal
 */
// so that we can test easier
export function __requireModule(localRequire: typeof requireModule, localResolve: typeof resolveModule) {
  requireModule = localRequire
  resolveModule = localResolve
}
