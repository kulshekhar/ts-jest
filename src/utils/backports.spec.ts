import { inspect } from 'util'

import { testing } from 'bs-logger'
import set from 'lodash/set'

import { backportJestConfig } from './backports'

const logger = testing.createLoggerMock()
const logTarget = logger.target

beforeEach(() => {
  logTarget.clear()
})

describe('backportJestConfig', () => {
  const makeTestsFor = (oldPath: string, values: unknown[]) => {
    values.forEach((val) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      let original: any

      beforeEach(() => {
        original = {}
        set(original, oldPath, val)
      })

      describe(`with "${oldPath}" set to ${inspect(val)}`, () => {
        it('should warn the user', () => {
          backportJestConfig(logger, original)

          expect(logTarget.lines.warn).toMatchSnapshot()
        }) // should warn the user

        it('should have changed the config correctly', () => {
          expect(original).toMatchSnapshot('before')
          expect(backportJestConfig(logger, original)).toMatchSnapshot('migrated')
        }) // should have changed the config
      }) // with xxx set to yyy
    }) // for
  } // makeTestsFor

  makeTestsFor('globals.__TS_CONFIG__', [{ foo: 'bar' }])

  makeTestsFor('globals.__TRANSFORM_HTML__', [true, false])

  makeTestsFor('globals.ts-jest.tsConfigFile', ['tsconfig.build.json'])

  makeTestsFor('globals.ts-jest.tsConfig', ['tsconfig.build.json'])

  makeTestsFor('globals.ts-jest.enableTsDiagnostics', [true, false, '\\.spec\\.ts$'])

  makeTestsFor('globals.ts-jest.useBabelrc', [true, false])

  makeTestsFor('globals.ts-jest.typeCheck', [true, false])

  makeTestsFor('globals.ts-jest.skipBabel', [true, false])
})
