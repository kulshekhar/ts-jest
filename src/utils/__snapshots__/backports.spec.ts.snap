// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to false should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "__TRANSFORM_HTML__": false,
  },
}
`;

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to false should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {},
  },
}
`;

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to false should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.__TRANSFORM_HTML__\\" is deprecated, use \\"[jest-config].globals.ts-jest.stringifyContentPathRegex\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to true should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "__TRANSFORM_HTML__": true,
  },
}
`;

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to true should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "stringifyContentPathRegex": "\\\\.html?$",
    },
  },
}
`;

exports[`backportJestConfig with "globals.__TRANSFORM_HTML__" set to true should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.__TRANSFORM_HTML__\\" is deprecated, use \\"[jest-config].globals.ts-jest.stringifyContentPathRegex\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.__TS_CONFIG__" set to { foo: 'bar' } should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "__TS_CONFIG__": Object {
      "foo": "bar",
    },
  },
}
`;

exports[`backportJestConfig with "globals.__TS_CONFIG__" set to { foo: 'bar' } should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "tsconfig": Object {
        "foo": "bar",
      },
    },
  },
}
`;

exports[`backportJestConfig with "globals.__TS_CONFIG__" set to { foo: 'bar' } should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.__TS_CONFIG__\\" is deprecated, use \\"[jest-config].globals.ts-jest.tsconfig\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to '\\\\.spec\\\\.ts$' should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "enableTsDiagnostics": "\\\\.spec\\\\.ts$",
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to '\\\\.spec\\\\.ts$' should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "diagnostics": Object {
        "pathRegex": "\\\\.spec\\\\.ts$",
        "warnOnly": true,
      },
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to '\\\\.spec\\\\.ts$' should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.enableTsDiagnostics\\" is deprecated, use \\"[jest-config].globals.ts-jest.diagnostics\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to false should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "enableTsDiagnostics": false,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to false should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "diagnostics": false,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to false should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.enableTsDiagnostics\\" is deprecated, use \\"[jest-config].globals.ts-jest.diagnostics\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to true should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "enableTsDiagnostics": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to true should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "diagnostics": Object {
        "warnOnly": true,
      },
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.enableTsDiagnostics" set to true should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.enableTsDiagnostics\\" is deprecated, use \\"[jest-config].globals.ts-jest.diagnostics\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to false should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "skipBabel": false,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to false should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "babelConfig": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to false should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.skipBabel\\" is deprecated, use \\"[jest-config].globals.ts-jest.babelConfig\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to true should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "skipBabel": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to true should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {},
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.skipBabel" set to true should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.skipBabel\\" is deprecated, use \\"[jest-config].globals.ts-jest.babelConfig\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.tsConfigFile" set to 'tsconfig.build.json' should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "tsConfigFile": "tsconfig.build.json",
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.tsConfigFile" set to 'tsconfig.build.json' should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "tsconfig": "tsconfig.build.json",
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.tsConfigFile" set to 'tsconfig.build.json' should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.tsConfigFile\\" is deprecated, use \\"[jest-config].globals.ts-jest.tsconfig\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.typeCheck" set to false should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "typeCheck": false,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.typeCheck" set to false should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "isolatedModules": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.typeCheck" set to false should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.typeCheck\\" is deprecated, use \\"[jest-config].globals.ts-jest.isolatedModules\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.typeCheck" set to true should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "typeCheck": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.typeCheck" set to true should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "isolatedModules": false,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.typeCheck" set to true should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.typeCheck\\" is deprecated, use \\"[jest-config].globals.ts-jest.isolatedModules\\" instead.
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to false should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "useBabelrc": false,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to false should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "babelConfig": Object {},
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to false should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.useBabelrc\\" is deprecated, use \\"[jest-config].globals.ts-jest.babelConfig\\" instead.
    ↳ See \`babel-jest\` related issue: https://github.com/facebook/jest/issues/3845
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to true should have changed the config correctly: before 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "useBabelrc": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to true should have changed the config correctly: migrated 1`] = `
Object {
  "globals": Object {
    "ts-jest": Object {
      "babelConfig": true,
    },
  },
}
`;

exports[`backportJestConfig with "globals.ts-jest.useBabelrc" set to true should warn the user 1`] = `
Array [
  "[level:40] \\"[jest-config].globals.ts-jest.useBabelrc\\" is deprecated, use \\"[jest-config].globals.ts-jest.babelConfig\\" instead.
    ↳ See \`babel-jest\` related issue: https://github.com/facebook/jest/issues/3845
",
  "[level:40] Your Jest configuration is outdated. Use the CLI to help migrating it: ts-jest config:migrate <config-file>.
",
]
`;
